File: go/98_Validate_Binary_Search_Tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isValidBST(root *TreeNode) bool {
	return helper(root, math.MinInt64, math.MaxInt64)
}

func helper(root *TreeNode, low int, upper int) bool {
	if root == nil {
		return true
	}
	if root.Val <= low || root.Val >= upper {
		return false
	}
	return helper(root.Left, low, root.Val) && helper(root.Right, root.Val, upper)
}

func isValidBST2(root *TreeNode) bool {
	stack := []*TreeNode{}
	inorder := math.MinInt64

	for len(stack) > 0 || root != nil {
		for root != nil {
			stack = append(stack, root)
			root = root.Left
		}
		root = stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		if root.Val <= inorder {
			return false
		}
		inorder = root.Val
		root = root.Right
	}
	retrun true
}

File: python/98_Validate_Binary_Search_Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        def helper(root,low = float('-inf'),upper = float('inf')):
            if not root:
                return True
            if root.val <= low or root.val >= upper:
                return False
            return helper(root.left, low,root.val) and helper(root.right, root.val,upper)
        return helper(root)

    def isValidBST2(self, root: TreeNode) -> bool:
        stack = []
        inorder = float('-inf')
        
        while stack or root:
            while root:
                stack.append(root)
                root = root.left
            root = stack.pop()
            if root.val <= inorder:
                return False
            inorder = root.val
            root = root.right

        return True